core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -156,10 +156,7 @@ public void commit() {
         .onlyRetryOn(CommitFailedException.class)
         .run(item -> {
           TableMetadata updated = internalApply();
-          // only commit the updated metadata if at least one snapshot was removed
-          if (updated.snapshots().size() != base.snapshots().size()) {
-            ops.commit(base, updated);
-          }
```

Comment:
How about also updating `internalApply` so that it does this check?

This is intended to only modify the table and run a commit if the table changes. But you're right that when combined with a transaction, this will cause a failure because commit is not called at all.

But we still want to avoid unnecessary commits. To do that, `internalApply` should do this check on the new `TableMetadata` it produces. If the new metadata has the same set of manifests as the original, then it should return the original. That will cause the commit to be skipped in most cases because the actual commit logic inside of `ops.commit` checks whether the metadata is the same as the base metadata and returns if it is.

Reply:
Thanks for reply, I know your point, in most transaction cases in can commit, but in BaseMetastoreTableOperations, we can return a original then it will return without any commit. I will update the code and add more UTs

Reply:
I think if we return the original one, then the logic inside of TransactionTableOperations will put the same metadata to current, like : BaseTransaction.this.current = metadata;  this will also cause failure in commitTransaction because in commitTransaction it has to check whether lastBase != current.

Reply:
We should definitely not copy the data. The purpose of returning the original metadata is to avoid an unnecessary commit if the metadata hasn't changed. The `TableOperations` objects use identity to determine if anything has changed.

There may be a check inside of `BaseTransaction` that also needs to be updated. But the right way to handle that is to update that check to allow passing the same metadata back.

Reply:
I have tried to change a preCondition check to be a if-else block, this may work but we may need to change some UTs which are assert to return a IllegalStateException.

Reply:
@rdblue I have change the behavior for commitTransaction, make it the same as what commit does in TableOperations. This will avoid to copy the data and also allow to pass the same metadata.